osu.Game.Rulesets.Osu/Difficulty/Skills/ProbabilityBasedSkill.cs

```diff
@@ -0,0 +1,410 @@
+                if (sectionPerformanceData.ExpectedDurationOfFailedFullComboAttempts < easiestSectionPerformanceData.ExpectedDurationOfFailedFullComboAttempts)
+                    easiestSectionPerformanceData = sectionPerformanceData;
+            }
+
+            return easiestSectionPerformanceData;
+        }
+
+        private MapSectionPerformanceData[] getPerformanceDataForSections(double skill)
+        {
+            var result = new MapSectionPerformanceData[difficulty_count];
+
+            for (int i = 0; i < difficulty_count; i++)
+            {
+                int sectionStartIndex = noteDifficulties.Count * i / difficulty_count;
+                int sectionEndIndex = noteDifficulties.Count * (i + 1) / difficulty_count - 1;
+
+                result[i] = getPerformanceDataForSection(skill, sectionStartIndex, sectionEndIndex);
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Calculate the average time a player with the given skill will take to Full Combo the given section of a map.
+        /// </summary>
+        /// <param name="skill">The skill level of the player.</param>
+        /// <param name="first">Index of the first object in the section.</param>
+        /// <param name="last">Index of the last object in the section.</param>
+        private MapSectionPerformanceData getPerformanceDataForSection(double skill, int first, int last)
+        {
+            double powSkill = Math.Pow(skill, -difficultyExponent);
+            double powDifficulty = exponentiatedMapSectionDifficulty(noteDifficulties[first], noteDifficulties[last]);
+
+            var result = new MapSectionPerformanceData
+            {
+                StartTime = noteDifficulties[first].PreviousStartTime,
+                EndTime = noteDifficulties[last].StartTime,
+                ExpectedTimeUntilFullCombo = 0,
+                ExponentiatedDifficulty = powDifficulty,
+                FullComboProbability = fcProbabilityPrecomputed(powSkill, powDifficulty)
+            };
+
+            for (int i = first; i <= last; ++i)
+            {
+                var note = noteDifficulties[i];
+                double hitProbability = fcProbabilityPrecomputed(powSkill, note.ExponentiatedDifficulty) + 1e-10;
+                result.ExpectedTimeUntilFullCombo = (result.ExpectedTimeUntilFullCombo + note.DeltaTime) / hitProbability;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// The estimated time it takes a player to full-combo a particular submap,
+        /// assuming they retry as soon as they miss a note.
+        /// </summary>
+        private MapSectionPerformanceData aggregateMapSectionPerformanceData(MapSectionPerformanceData[] sectionData, int first, int count)
+        {
+            int last = first + count - 1;
+
+            // TODO: clarify - is this just the approximate time needed to restart the map?
```

Comment:
Should be removed (from my understanding this was indeed the approximate time that a user takes to restart a map).